### PR TITLE
blockchain: Generalize and optimize chain reorg.

### PR DESCRIPTION
**Pull request by davecgh**
This generalizes and optimizes the reorganize chain function to take advantage of the linear structure of the nodes being attached and detached in order to avoid reloading the parent blocks as well as allow it to rewind the chain in the case no attach nodes are provided or extend the chain in the case no detach nodes are provided.

It also adds several assertions to ensure the assumptions about the state hold and generally cleans up the function for code consistency.

### CODE REVIEW

File: blockchain/chain.go [hunk 1]
```diff
@@ -1286,8 +1346,11 @@ func (b *BlockChain) reorganizeChain(detachNodes, attachNodes *list.List, flags
 	// at least a couple of ways accomplish that rollback, but both involve
 	// tweaking the chain and/or database.  This approach catches these
 	// issues before ever modifying the chain.
-	var topBlock *blockNode
-	for e := attachNodes.Front(); e != nil; e = e.Next() {
+	for i, e := 0, attachNodes.Front(); e != nil; i, e = i+1, e.Next() {
```
Comment (alexlyp) [inline blockchain/chain.go]: Why is i necessary here?

Comment (davecgh) [reply blockchain/chain.go]: Nice catch.  It's actually used in the next PR.  I originally wrote the whole series together, but then split them up for easier review.  I'll move it to the proper commit that uses it.

Comment (alexlyp) [reply blockchain/chain.go]: Gotcha, thanks for the clarification.

Comment (davecgh) [reply blockchain/chain.go]: Done

**Pull request merged**